design-decisions/2018-09-17-Builtin-Map.md

```diff
@@ -0,0 +1,71 @@
+Builtin Maps
+============
+
+Background
+----------
+
+K provides a builtin `Map` sort representing partial maps on sort `K`.
+Here "partial" means that some keys may not be associated with any value.
+KEVM relies on concrete operations only, but we will eventually need to support
+symbolic operations.
+
+Operations
+----------
+
+The builtin `Map` domain should support operations equivalent to:
+
+```
+module MAP
+    sort Map{} []
+
+    // unit: The empty map (no keys).
+    symbol unit{}() : Map{} []
+
+    // element(key, value): A map from one key to one value.
+    symbol element{}(K{}, K{}) : Map{} []
+
+    // concat(map1, map2): Combine both maps so that a key is in the result if
+    // it is in either operand.
+    symbol concat{}(Map{}, Map{}) : Map{} []
+
+    // update(map, key, value): Update the map so that key refers to value.
+    symbol update{}(Map{}, K{}, K{}) : Map{} []
+
+    // lookup(map, key): Lookup key in map and return the associated value.
+    // Return \bottom if the key is missing.
+    symbol lookup{}(Map{}, K{}) : K{} []
+
+    // inKeys(map, key): Is key in map?
+    symbol inKeys{}(Map{}, K{}) : Bool{} []
+endmodule
+```
+
+Questions:
+
+- What does `update` do if the key is missing from the map?
+- What does `concat` do if the key is present in both maps?
```

Comment:
@ttuegel @traiansf I think `concat(m1,m2)` shouldn't be defined when there is a key in both `m1` and `m2`, like the separating conjunction `*` operator.